bug 1255609 - Fixes for spam dashboard

jwhitlock opened the pull request:
Fix issues with PR #3816:

* On the "dashboard" (the Django admin list view), the Document name is length-restricted, so that long names do not cause the list view to become overly long and require scrolling to get to the review dropdown.
* An email is sent to the "first edits" email list when an edit is blocked, but not again when the block is reviewed.

Also improves ``make coveragetest``, to generate the HTML coverage docs and to recreate the skeleton of the 'cleaned' ``build`` folder.

jwhitlock commented:
Pre-emptive force-push to change double quotes to single quotes.

jezdez commented on kuma/wiki/admin.py:
```diff
@@ -363,6 +364,30 @@ def slug_short(self, obj):
+        """
+        doc = obj.document
+        if doc:
+            full_path = '/%s/docs/%s' % (doc.locale, doc.slug)
```
The slug may contain non-ASCII characters so this needs to be a Unicode string

jwhitlock replied:
I think I'm going to lean on ``Document.get_absolute_url()`` to do the work. I'm having trouble finding good unicode URLs for testing - https://developer.mozilla.org/fr/docs/Accessibilité/ARIA/Techniques_ARIA/Utiliser_le_role_progressbar is closest.

jwhitlock replied:
Nope, ``Document.get_absolute_url()`` appears to be broken on Bengali, like on https://developer.mozilla.org/bn-BD/docs/Web/Guide/HTML/HTML5_ডকুমেন্টের_সেকশন_এবং_আউটলাইন.  Manually constructing the path from the slug.

jezdez commented on kuma/wiki/admin.py:
```diff
@@ -363,6 +364,30 @@ def slug_short(self, obj):
+        doc = obj.document
+        if doc:
+            full_path = '/%s/docs/%s' % (doc.locale, doc.slug)
+            if len(full_path) <= 25:
```
If you want to have consistent length between non-ASCII characters using `len(unicodedata.normalize(full_path, 'NFC'))` would help, assuming `full_path` is a Unicode string

jezdez commented on kuma/wiki/admin.py:
```diff
@@ -363,6 +364,30 @@ def slug_short(self, obj):
+
+        The important part is having an HTML break character such as a space,
+        so truncating paths to 25 characters after the locale, and long titles
+        at 25 characters.
```
The 25 characters would make sense to store in a parameter or variable so it's easier to see what's going on.

jezdez commented on kuma/wiki/admin.py:
```diff
@@ -363,6 +364,30 @@ def slug_short(self, obj):
+            else:
+                # 25 characters like /en_US/...End/Of/Slug
+                slash2_pos = full_path.index('/', 1) + 1
+                path = '%s...%s' % (full_path[:slash2_pos],
```
Since we can use Unicode here, I'd guess it would help to use an Ellipsis here, `…`.

jezdez commented on kuma/wiki/admin.py:
```diff
@@ -363,6 +364,30 @@ def slug_short(self, obj):
+                path = '%s...%s' % (full_path[:slash2_pos],
+                                    full_path[(slash2_pos - 22):])
+            title = truncatechars(doc.title, 25)
+            return '%s (%s)' % (path, title)
```
Needs to be Unicode.

jezdez commented on kuma/wiki/admin.py:
```diff
@@ -363,6 +364,30 @@ def slug_short(self, obj):
+            return '%s (%s)' % (path, title)
+        else:
+            return mark_safe('<em>new document</em>')
+    doc_short.short_description = 'Document (if edit)'
```
Not sure if needed, but marking this for translation is not a bad idea

jwhitlock replied:
This would be the first translated string (from kuma) in the Django admin in the wiki app, if not over the whole project. Is now the time to cross that line?

jezdez replied:
Good point, I think there is a ticket about that IIRC. Carry on :+1:

jwhitlock replied:
I can't find an existing ticket. I also think we shouldn't localize the admin, so I'm not opening one.

jezdez commented on kuma/wiki/apps.py:
```diff
@@ -137,7 +139,11 @@ def on_revision_save(self, sender, instance, **kwargs):
+            self, sender, instance, created, raw, **kwargs):
+        if raw or not created:
+            # Only send for new instances, not fixtures or edits
+            return
```
Ah, good idea

jezdez commented on kuma/wiki/apps.py:
```diff
@@ -146,4 +152,4 @@ def on_document_spam_attempt_save(self, sender, instance, **kwargs):
                                 {'spam_attempt': instance})
         send_mail(subject, body, settings.DEFAULT_FROM_EMAIL,
-                  ['mdn-spam-watch@mozilla.com'])
+                  [config.EMAIL_LIST_FOR_FIRST_EDITS])
```
Huh, didn't see that, I wonder if it would make sense to rename the config variable to better express what this is used for (e.g. `EMAIL_LIST_SPAM_WATCH`).

jwhitlock replied:
I like the name ``EMAIL_LIST_SPAM_WATCH`` and using it for both. How do we ship that change without breaking production?

jezdez replied:
Check in the prod constance admin if the default of the config value has been changed since it was deployed and change the value of the renamed config variable once deployed.

jwhitlock replied:
The default and current value of ``EMAIL_LIST_FOR_FIRST_EDITS`` is ``mdn-spam-watch@mozilla.com``. Adding another commit to make the switch to ``EMAIL_LIST_SPAM_WATCH``...

jezdez commented on kuma/wiki/admin.py:
```diff
@@ -363,6 +364,30 @@ def slug_short(self, obj):
+            if len(full_path) <= 25:
+                path = full_path
+            else:
+                # 25 characters like /en_US/...End/Of/Slug
```
Locales are written as `en-US` in URLs

jezdez commented on kuma/wiki/tests/test_admin.py:
```diff
@@ -54,6 +55,30 @@ def test_slug_short(self):
+        slug = 'NotSpam'
+        html = '<p>This page is not spam.</p>'
+        doc = document(title='blah', slug=slug, html=html, save=True,
+                       locale='en_US')
```
I know the test will pass nevertheless, but to prevent red herrings in locale parsing, I'd use `en-US` here

jwhitlock replied:
Did you know I have trouble with "-" vs "_"? It's true!

Changing to not set the locale, so we get the default ``en-US``, and changing the tests accordingly.

jezdez commented on Makefile:
```diff
@@ -8,7 +8,7 @@ test:
 
 coveragetest: clean
-	py.test --cov=$(target) $(target)
+	py.test --cov=$(target) --cov-report term --cov-report html $(target)
```
Is generating the html report needed here?

jwhitlock replied:
"term" is the default, which I always follow with "coverage html" to generate the HTML report, which include annotations of the missed lines and branches. It is unneeded in TravisCI, though.

jezdez commented:
r+wc

jwhitlock commented:
Force-push to rebase on master, and:

* Instead of changing ``coveragetest`` target to generate HTML, added a ``coveragetesthtml`` target
* Add a class constant ``DocumentSpamAttemptAdmin.MAX_LENGTH`` instead of using 25 everywhere
* Change to standard end-truncation of slugs in ``doc_short`` instead of funky middle truncation, which still didn't really work for long paths and titles.
* Better unicode handling with explicit ``u'Unicode format %s'``, ``…`` ellipsis character instead of three periods, and Django's unicode-aware ``Truncator``
* Split ``doc_short`` tests instead of serially testing different slugs and titles.

jezdez commented on kuma/wiki/tests/test_admin.py:
```diff
@@ -54,6 +56,41 @@ def test_slug_short(self):
+                       locale='bn-BD')
+        revision(document=doc, content=html, is_approved=True, save=True)
+        dsa = DocumentSpamAttempt(slug=slug, document=doc)
+        expected = u'/bn-BD/docs/Web/Guide/HT… (HTML5 ডকুমেন্টের সেকশন এব…)'
```
Nice!

jezdez commented:
@jwhitlock Awesome changes!

jwhitlock commented:
One more commit to standardize on ``EMAIL_LIST_SPAM_WATCH``. @jezdez, if this looks good feel free to merge.

jezdez commented:
:boom:
